ndb/src/google/cloud/ndb/model.py

```diff
@@ -2947,28 +2962,286 @@ def __init__(self, *args, **kwargs):
+    def _get_arg(cls, kwargs, keyword):
+        """Parse keywords for fields that aren't user-defined properties.
+
+        This is used to re-map special keyword arguments in the presence
+        of name collision. For example if ``id`` is a property on the current
+        :class:`Model`, then it may be desirable to pass ``_id`` (instead of
+        ``id``) to the constructor.
+
+        If the argument is found as ``_{keyword}`` or ``{keyword}``, it will
+        be removed from ``kwargs``.
+
+        Args:
+            kwargs (Dict[str, Any]): A keyword arguments dictionary.
+            keyword (str): A keyword to be converted.
+
+        Returns:
+            Optional[Any]: The ``keyword`` argument, if found.
+        """
+        alt_keyword = "_" + keyword
+        if alt_keyword in kwargs:
+            return kwargs.pop(alt_keyword)
+
+        if keyword in kwargs:
+            obj = getattr(cls, keyword, None)
+            if not isinstance(obj, Property) or isinstance(obj, ModelKey):
+                return kwargs.pop(keyword)
+
+        return None
+
+    def _set_attributes(self, kwargs):
+        """Set attributes from keyword arguments.
+
+        Args:
+            kwargs (Dict[str, Any]): A keyword arguments dictionary.
+        """
+        cls = type(self)
+        for name, value in kwargs.items():
+            # NOTE: This raises an ``AttributeError`` for unknown properties
+            #       and that is the intended behavior.
+            prop = getattr(cls, name)
+            if not isinstance(prop, Property):
+                raise TypeError("Cannot set non-property {}".format(name))
+            prop._set_value(self, value)
 
     @classmethod
     def _get_kind(cls):
-        """Return the kind name for this class.
+        """str: Return the kind name for this class.
 
         This defaults to ``cls.__name__``; users may override this to give a
         class a different name when stored in Google Cloud Datastore than the
         name of the class.
         """
         return cls.__name__
 
+    def _set_projection(self, projection):
+        """Set the projected properties for this instance.
+
+        Args:
+            projection (Union[list, tuple]): An iterable of strings
+                representing the projection for the model instance.
+        """
+        self._projection = tuple(projection)
+
+    @classmethod
+    def _fix_up_properties(cls):
+        """Fix up the properties by calling their ``_fix_up()`` method.
+
+        .. note::
+
+            This is called by :class:`MetaModel`, but may also be called
+            manually after dynamically updating a model class.
+
+        Raises:
+            KindError: If the returned kind from ``_get_kind()`` is not a
+                :class:`str`.
+            TypeError: If a property on this model has a name beginning with
+                an underscore.
+        """
+        kind = cls._get_kind()
+        if not isinstance(kind, str):
```

Comment:
Previously this did a `basestring` check, but I think restricting to `str` (i.e. not allowing `bytes`) is the right move in Python 3. @chrisrossi shall I put a comment in `MIGRATION_NOTES.md` or nah?

Reply:
I don't think it's strictly necessary. You can if you want.